### ISSUE TITLE
RooFit backend library: force to specific architecture

### ISSUE THREAD

**Issue by Axel-Naumann** [root-project/root#7361]
### Is your feature request related to a problem? Please describe.

RooFit compute libraries are loaded at runtime, picking the best instruction set available. Some of our users (LHCb, for instance), need to pin down the specific library to be loaded, to make sure their results are specifically the same as the ones they validated.

We need to
- [ ] have a mechanism for this
- [ ] expose the mechanism, for instance through documentation / release notes
- [ ] ensure that LHCb gets that flag enabled.

This *must* happen by v6.24 for this release to be usable by LHCb.

### Describe the solution you'd like

Environment variable: `export ROOT_ARCH=avx2` etc.

### Describe alternatives you've considered

Preventing all LCG builds from using optimal libraries seems like overkill, so this should not be a cmake flag.

### Additional context

Calling in @pikacic to share additional constraints / comments / ideas!

**Comment by pikacic** [comment]
Hi,

the LHCb point of view on the issue is that we must be able to know which code we executed, mostly for debugging purposes (understanding why the trigger application took a specific decision counts as debugging for me).

When you have an application that can decide which version of a function to run depending on something you cannot control, like the specific CPU on the Grid node or the time of the day, makes the debugging much more tricky (at some point we started storing in each event the binary id and the CPU type used for that event), so we decided that we do not want this kind of automatic dispatching: an application compiled to run on a Nehalem equivalent CPU should run only instructions valid for Nehalem processors, even if the actual CPU can do better.

Our approach is that we still have some functions that have different implementations for different instructions sets (basically scalar, sse4.2 and avx2), but we chose at compile time which one to use by adding something like `-mavx2` or `-march=haswell` to the compile flags. And yes, it should be a cmake flag, but you are most welcome to have also the `ARCH=dynamic` option.

OTOH you are talking about RooFit and AFAIK we do not use it in trigger, reconstruction or simulation software, but it is only used for the analysis, and there we have less constraints, as it has become common practice to use Conda based installations, so builds and versions decoupled from our production software.

**Comment by Axel-Naumann** [comment]
Thanks, @pikacic ! From this and your discussion with @oshadura our current understanding is that we can release v6.24/00 without a mechanism to force a specific architecture for the RooFit computation libraries.

It is so far unclear to me whether LCG will be able to provide what you are hoping to see (a pinning of the architecture at configuration time) without preventing this 10x acceleration from being accessible for instance to ATLAS physicists. Until this is decided we will provide a mechanism to pin the architecture through an environment variable.

**Comment by pikacic** [comment]
Your understanding is correct, you can go ahead with v6.24/00.

For the compile time pinning, from the discussion we had yesterday, we can have architecture specific builds, so that, for example, `x86_64-centos7-gcc9-opt` will stick to the basic `x86-64` instructions, while `haswell-centos7-gcc9-opt` will do its best using AVX2 and all other goodies you get on that platform (we can also use the microarchitecture levels like `x86-64-v2` which will be available as of GCC 11 and Clang 12).

For the RooFit specific case, our analysis does not use the builds from SFT and we do not use RooFit in trigger and reconstruction (which are the cases that I'm most worried about). Let's see from somebody in ATLAS what they think. I anyway (as a matter of principle) would prefer if you can have dynamic architecture dispatching that I can turn off in specific builds.

@oshadura also mention possible optimizations in compression algorithms, but, again, that's a case that does not worry me, as they only deal with integers, so you do not get incompatible/incorrect results by changing the order of operations (unless there's a bug).

The bottom line is that I'm worried about numerical instabilities due to the change of order of floating point operations in trigger and reconstruction code. It doesn't matter if scalar or avx2 versions produce slightly different results, as long as I know that running the scalar version of the application will always give me the same result.

**Comment by lmoneta** [comment]
This issue should have been closed after #7415 was merged
